scripts/saml-configuration

```diff
@@ -0,0 +1,144 @@
+except Exception:
+    logging.critical('Cannot receive Zulip SAML configuration', exc_info=True)
+    sys.exit(1)
+
+for idp in idps:
+    # The list of the IdPs signing certificates.
+    certs_list: List[str] = []
+    new_cert: str
+    new_cert_object: x509.Certificate
+
+    # Check the provided settings.
+    if 'metadata_url' not in idps[idp] or 'entity_id' not in idps[idp]:
+        logging.error('No \'metadata_url\' and/or \'entity_id\' field for IdP '
+                      '\'%s\' in the SAML configuration', idp)
+        continue
+
+    # Get the data from the metadata url of the IdP.
+    data: Dict[str, Any] = Parser.parse_remote(
+        idps[idp]['metadata_url'],
+        validate_cert = True,
+        entity_id = idps[idp]['entity_id']
+    )
+
+    if 'idp' not in data:
+        logging.error('Cannot receive metadata for IdP %s', idp)
+        logging.debug(data)
+        continue
+
+    # Some documentation:
+    # - https://github.com/onelogin/python3-saml
+    # - https://github.com/onelogin/python3-saml/blob/master/src\
+    #     /onelogin/saml2/idp_metadata_parser.py
+
+    if 'x509cert' in data['idp']:
+        # A single certificate, used for both signing and encryption.
+        certs_list.append(data['idp']['x509cert'])
+    elif ('x509certMulti' in data['idp']
+          and 'signing' in data['idp']['x509certMulti']):
+        # There are multiple certificates, maybe different ones for
+        # signing and encryption.
+        certs_list.extend(data['idp']['x509certMulti']['signing'])
+
+    if not certs_list:
+        logging.error('Cannot find any certificates for IdP %s', idp)
+        continue
+
+    # Format the certificates properly.
+    # See https://github.com/onelogin/python3-saml/blob/master/src\
+    #   /onelogin/saml2/utils.py
+    certs: Iterator[str] = map(Utils.format_cert, certs_list)
+
+    # Map the certificates to their x509 objects in order to be able to
+    # determine the validity of the certificate.
+    # See https://cryptography.io/en/latest/x509/reference.html\
+    #   #cryptography.x509.Certificate
+    certs_objects: Iterator[Tuple[str, x509.Certificate]] = (
+        (cert, x509.load_pem_x509_certificate(cert.encode(), backend=default_backend()))
+        for cert in certs
+    )
+
+    # Check for the best certificate available:
+    #  - It is currently valid.
+    #  - It is valid for as long as possible.
+    try:
+        new_cert, new_cert_object = max(
+            (
+                (cert, cert_object)
+                for cert, cert_object in certs_objects
+                if is_valid(cert_object)
+            ),
+            key=lambda t: t[1].not_valid_after,
+        )
+    except ValueError:
+        logging.error('Cannot find any valid certificate for IdP %s', idp)
+        logging.debug('', exc_info=True)
+        continue
+
+    # Write the certificate to the appropriate location.
+    with open(idp_cert_path_template.format(idp), mode = 'w') as f:
```

Comment:
will this file have correct permissions?

Reply:
It should have `-rw-rw-r--` per default, but you're right, I'll change this to `-rw-------`.